velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp

```diff
@@ -673,6 +679,345 @@ TEST_F(DateTimeFunctionsTest, dateTruncDate) {
 }
 
+TEST_F(DateTimeFunctionsTest, dateAddDate) {
+  const auto dateAdd = [&](const std::string& unit,
+                           std::optional<int32_t> value,
+                           std::optional<Date> date) {
+    return evaluateOnce<Date>(
+        fmt::format("date_add('{}', c0, c1)", unit), value, date);
+  };
+
+  const auto parseDate = [&](const std::string& strDate) -> Date {
+    Date result;
+    parseTo(strDate, result);
+    return result;
+  };
+
+  // Check null behaviors
+  EXPECT_EQ(std::nullopt, dateAdd("day", 1, std::nullopt));
+  EXPECT_EQ(std::nullopt, dateAdd("month", std::nullopt, Date(0)));
+
+  // Check invalid units
+  EXPECT_THROW(dateAdd("millisecond", 1, Date(0)), VeloxUserError);
+  EXPECT_THROW(dateAdd("second", 1, Date(0)), VeloxUserError);
+  EXPECT_THROW(dateAdd("minute", 1, Date(0)), VeloxUserError);
+  EXPECT_THROW(dateAdd("hour", 1, Date(0)), VeloxUserError);
+  EXPECT_THROW(dateAdd("invalid_unit", 1, Date(0)), VeloxUserError);
+
+  // Simple tests
+  EXPECT_EQ(
+      parseDate("2019-03-01"), dateAdd("day", 1, parseDate("2019-02-28")));
+  EXPECT_EQ(
+      parseDate("2020-03-28"), dateAdd("month", 13, parseDate("2019-02-28")));
+  EXPECT_EQ(
+      parseDate("2020-02-28"), dateAdd("quarter", 4, parseDate("2019-02-28")));
+  EXPECT_EQ(
+      parseDate("2020-02-28"), dateAdd("year", 1, parseDate("2019-02-28")));
+
+  // Account for the last day of a year-month
+  EXPECT_EQ(
+      parseDate("2020-02-29"), dateAdd("day", 395, parseDate("2019-01-30")));
+  EXPECT_EQ(
+      parseDate("2020-02-29"), dateAdd("month", 13, parseDate("2019-01-30")));
+  EXPECT_EQ(
+      parseDate("2020-02-29"), dateAdd("quarter", 1, parseDate("2019-11-30")));
+  EXPECT_EQ(
+      parseDate("2030-02-28"), dateAdd("year", 10, parseDate("2020-02-29")));
+
+  // Check for negative intervals
+  EXPECT_EQ(
+      parseDate("2019-02-28"), dateAdd("day", -366, parseDate("2020-02-29")));
+  EXPECT_EQ(
+      parseDate("2019-02-28"), dateAdd("month", -12, parseDate("2020-02-29")));
+  EXPECT_EQ(
+      parseDate("2019-02-28"), dateAdd("quarter", -4, parseDate("2020-02-29")));
+  EXPECT_EQ(
+      parseDate("2018-02-28"), dateAdd("year", -2, parseDate("2020-02-29")));
+}
+
+TEST_F(DateTimeFunctionsTest, dateAddTimestamp) {
+  const auto dateAdd = [&](const std::string& unit,
+                           std::optional<int32_t> value,
+                           std::optional<Timestamp> timestamp) {
+    return evaluateOnce<Timestamp>(
+        fmt::format("date_add('{}', c0, c1)", unit), value, timestamp);
+  };
+
+  // Check null behaviors
+  EXPECT_EQ(std::nullopt, dateAdd("second", 1, std::nullopt));
+  EXPECT_EQ(std::nullopt, dateAdd("month", std::nullopt, Timestamp(0, 0)));
+
+  // Check invalid units
+  EXPECT_THROW(dateAdd("invalid_unit", 1, Timestamp(0, 0)), VeloxUserError);
+
+  // Simple tests
+  EXPECT_EQ(
+      Timestamp(1551348061, 999'999) /*2019-02-28 10:01:01.000*/,
```

Comment:
Is it hard to mock a `parseTimestamp()` function?

Reply:
Looking at it. Thank you for pointing out!